### PR TITLE
[All / Desktop] Add pinned notes feature

### LINKED ISSUE #296: [Feature Request] pinned notes
### Operating system

- [x] Windows
- [x] macOS
- [x] Linux
- [x] Android
- [x] iOS

### Application

- [x] Desktop
- [x] Mobile
- [x] Terminal

It could be great if we could pinned notes  and then display its title on top of all folders and tags in the menu at left.

That way we could access them without searching it nor following the path of folders or tags.

For example, I never have any loyalty card in my pocket, but I could find the loyalty card number to provide it when needed :)

### PR DESCRIPTION
**Pull request by anihm136**
Fixes #296 

Notes can be marked as pinned from the context menu in the note list. Once pinned, they are displayed at the top of the notebook, sorted among themselves in the same order as rest of the notes. Pinned notes have lower priority than uncompleted todos (if that option is being used) i.e, the following order is followed:
* Uncompleted pinned todos
* Uncompleted unpinned todos
* Completed pinned todos and pinned notes
* Completed unpinned todos and unpinned notes

Items can be marked as pinned in the desktop client only (through context menu), and pinned items are displayed in bold in the notelist of desktop client. However, sorting is platform-independent

ezgif com-optimize

### CODE REVIEW

**Comment by mic704b** [comment]
> Once pinned, they are displayed at the top of the notebook, sorted among themselves in the same order as rest of the notes

Should not pinned notes be in fixed order?

Also you might include image/s in your PR to show what it looks like.

File: ReactNativeClient/lib/models/Note.js [hunk 1]
```diff
@@ -318,6 +324,13 @@ class Note extends BaseItem {
 			}
 		}
 
+		const cond = options.conditions.slice();
+		cond.push('pinned = 1');
+		const tempOptions = Object.assign({}, options);
+		tempOptions.conditions = cond;
```
Comment (mic704b) [inline ReactNativeClient/lib/models/Note.js]: Can this be done nicely without `cond`?

File: ElectronClient/gui/utils/NoteListUtils.js [hunk 2]
```diff
@@ -126,9 +126,22 @@ class NoteListUtils {
...
 			);
 
+			if (noteIds.length <= 1) {
+				menu.append(
+					new MenuItem({
+						label: _('Pin Item'),
+						click: async () => {
+							const newNote = await Note.load(noteIds[0]);
+							newNote.pinned = !newNote.pinned;
+							await Note.save(newNote);
```
Comment (mic704b) [inline ElectronClient/gui/utils/NoteListUtils.js]: Does pinning a note change it's updated timestamp?  Is this intentional?

File: ElectronClient/gui/utils/NoteListUtils.js [hunk 3]
```diff
@@ -126,9 +126,22 @@ class NoteListUtils {
...
 						clipboard.writeText(links.join(' '));
 					},
-				})
+				}),
 			);
 
+			if (noteIds.length <= 1) {
+				menu.append(
+					new MenuItem({
+						label: _('Pin Item'),
```
Comment (mic704b) [inline ElectronClient/gui/utils/NoteListUtils.js]: "Pin note" to make it consistent with existing context menu items.

File: ElectronClient/gui/utils/NoteListUtils.js (hunk 2 above)
Comment (anihm136) [reply ElectronClient/gui/utils/NoteListUtils.js]: Is there some way to update only one field of the note without changing the updated time? It was not intented.

File: ReactNativeClient/lib/models/Note.js (hunk 1 above)
Comment (anihm136) [reply ReactNativeClient/lib/models/Note.js]: Do you want me to rewrite more concisely with the same logic, or change the logic?

File: ElectronClient/gui/utils/NoteListUtils.js [hunk 4]
```diff
@@ -32,7 +32,7 @@ class NoteListUtils {
 							noteIds: noteIds,
 						});
 					},
-				})
+				}),
```
[the remark this answers is not in the document]
Comment (anihm136) [reply ElectronClient/gui/utils/NoteListUtils.js]: This was due to eslint errors on commit. I ran the fixer and it added those.

**Comment by anihm136** [comment]
> Should not pinned notes be in fixed order?

Do you mean in the order they were pinned?

File: ReactNativeClient/lib/joplin-database.js [hunk 5]
```diff
@@ -31,6 +31,7 @@ CREATE TABLE notes (
 	is_todo INT NOT NULL DEFAULT 0,
 	todo_due INT NOT NULL DEFAULT 0,
 	todo_completed INT NOT NULL DEFAULT 0,
+	pinned INT NOT NULL DEFAULT 0,
```
[the remark this answers is not in the document]
Comment (anihm136) [reply ReactNativeClient/lib/joplin-database.js]: Where can I add the migrations?

**Comment by mic704b** [comment]
> Do you mean in the order they were pinned?

No, not in the time order.  In whatever order the user wants them.

Pinned notes are "pinned" and so I understand that to mean they will not move. With multiple pinned notes, a user will set them in a particular order, and expect them to stay that way.  I would not expect them to reorder in response to the sort order settings.

This is my opinion though, I'm not sure what the spec is for this feature.

File: ElectronClient/gui/utils/NoteListUtils.js (hunk 2 above)
Comment (anihm136) [reply ElectronClient/gui/utils/NoteListUtils.js]: Will setting autotimestamp to false be enough here?

**Comment by anihm136** [comment]
For the migration, I see that I need to write a migration script, place it in the migrations folder and include it in the Migrations model. Is that right?

**Comment by anihm136** [comment]
> For the migration, I see that I need to write a migration script, place it in the migrations folder and include it in the Migrations model. Is that all I need to do??
Also, what is the numbering scheme for the migrations?

Never mind, I found the instructions in the joplin-database file.

File: ElectronClient/gui/utils/NoteListUtils.js (hunk 3 above)
Comment (anihm136) [reply ElectronClient/gui/utils/NoteListUtils.js]: Changed it to "Pin / Unpin note" to allow toggle.

File: ReactNativeClient/lib/models/Note.js (hunk 1 above)
Comment (anihm136) [reply ReactNativeClient/lib/models/Note.js]: Removed `cond` from logic

File: ReactNativeClient/lib/models/Note.js [hunk 6]
```diff
@@ -343,7 +356,8 @@ class Note extends BaseItem {
 			if ('limit' in tempOptions) tempOptions.limit -= uncompletedTodos.length;
 			const theRest = await this.search(tempOptions);
 
-			return uncompletedTodos.concat(theRest);
+			// return uncompletedTodos.concat(theRest);
```
[the remark this answers is not in the document]
Comment (anihm136) [reply ReactNativeClient/lib/models/Note.js]: Done

File: ReactNativeClient/lib/models/Note.js [hunk 7]
```diff
@@ -210,6 +210,10 @@ class Note extends BaseItem {
 			return uncompletedTodosOnTop && note.is_todo && !note.todo_completed;
 		};
 
+		const notePinned = note => {
+			return note.pinned;
```
[the remark this answers is not in the document]
Comment (anihm136) [reply ReactNativeClient/lib/models/Note.js]: Done

File: ReactNativeClient/lib/models/Note.js [hunk 8]
```diff
@@ -210,6 +210,10 @@ class Note extends BaseItem {
 			return uncompletedTodosOnTop && note.is_todo && !note.todo_completed;
 		};
 
+		const notePinned = note => {
```
[the remark this answers is not in the document]
Comment (anihm136) [reply ReactNativeClient/lib/models/Note.js]: Done

File: ReactNativeClient/lib/joplin-database.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (anihm136) [reply ReactNativeClient/lib/joplin-database.js]: Done

File: ElectronClient/gui/utils/NoteListUtils.js [hunk 9]
```diff
@@ -126,9 +126,22 @@ class NoteListUtils {
 						}
 						clipboard.writeText(links.join(' '));
 					},
-				})
+				}),
 			);
 
+			if (noteIds.length <= 1) {
```
[the remark this answers is not in the document]
Comment (anihm136) [reply ElectronClient/gui/utils/NoteListUtils.js]: Changed the label to allow pin and unpin.

**Comment by anihm136** [comment]
The latest changes are failing the tests for global search. I'm not sure why. The search engine is returning the right results, but its not reaching the note list. I'm working on fixing it now.

**Comment by anihm136** [comment]
It's passing all tests but one. I'm not sure which of my changes had anything to do with it, but it is failing for field-restricted search queries with chinese characters. How can I approach this?
Full error message: 
```
Failures:
1) services_SearchEngine should support field restricted queries with Chinese characters
  Message:
    Expected 0 to be 1, 'Test has thrown an exception - see above error'.
  Stack:
    Error: Expected 0 to be 1, 'Test has thrown an exception - see above error'.
        at <Jasmine>
        at UserContext.<anonymous> (/home/anirudh/Projects/forks/joplin/CliClient/tests-build/test-utils.js:381:15)
        at processTicksAndRejections (internal/process/task_queues.js:97:5)

264 specs, 1 failure
Finished in 585.743 seconds
Randomized with seed 93512 (jasmine --random=true --seed=93512)
```

**Comment by anihm136** [comment]
I've come up with a solution, but I feel it's more of a hack. The issue is that, when search is in a symbol based language (basicSearch), the sql parameters are being pushed to the array multiple times i.e, there is one variable to fill but 2 values given. I was unable to figure out what the root cause of it was, so I've added a check in the search function of the note.

**Comment by anihm136** [comment]
@mic704b not to be intrusive, but I just wanted to inform you that I have resolved most of the changes you had asked.
 The one doubt I have left is how to update the pinned attribute of the note without updating the timestamp. Will using `autotimestamp:false` work, or do I need to find another way?

File: ElectronClient/gui/utils/NoteListUtils.js (hunk 3 above)
Comment (mic704b) [reply ElectronClient/gui/utils/NoteListUtils.js]: Is it possible to switch it on context?

**Comment by mic704b** [comment]
> Will using `autotimestamp:false` work, or do I need to find another way?

It sounds right, but you need to check.

>  but I just wanted to inform you that I have resolved most of the changes you had asked

Have you added tests.  Automated tests where possible, and manual tests for things that can't be tested automatically.

**Comment by laurent22** [comment]
Not sure I want to merge this. We should have a concept of "favourite" notes in Joplin, which would show up somewhere (maybe on the sidebar), but having some notes ordered on top like this I'm not so keen.

**Comment by mic704b** [comment]
> Not sure I want to merge this

Understood.

@anihm136 if you want to continue with this, perhaps you can come up with a new proposal, and propose a spec for it before implementing.  In fact, this should probably be done for all new features.

**Comment by anihm136** [comment]
Alright. If it's not already there, I'll open a topic on the forum to discuss the "favorites" feature and hopefully get a spec. I'll keep this branch around on my fork in case it's ever needed.

**Pull request closed without merging**